### PR TITLE
alter columns

### PR DESCRIPTION
**Pull request by goodwanghan**

### CODE REVIEW

**Comment by goodwanghan** [comment]
not done, just finished the dataframe level implementation

**Comment by goodwanghan** [comment]
Tests to look at:
dataframe level change:

https://github.com/fugue-project/fugue/pull/113/files#diff-804a2e96dbe27be3b6cd4e23a9142d1d78c28877766240da7a712fbe1ddf1c72R224

workflow level change:

https://github.com/fugue-project/fugue/pull/113/files#diff-6d8345ac7d767c63f049bbd8dd9aac287a39afa9405c503c96f594671f930f8fR908

sql level change:

https://github.com/fugue-project/fugue/pull/113/files#diff-bb233c3f4a890eb6015253526359607a555e554209bc852e571405998c1680f2R494

File: fugue/dataframe/pandas_dataframe.py [hunk 1]
```diff
@@ -103,14 +106,38 @@ def _select_cols(self, cols: List[Any]) -> DataFrame:
...
+    def alter_columns(self, columns: Any) -> DataFrame:
+        new_schema = self._get_altered_schema(columns)
+        if new_schema == self.schema:
+            return self
+        new_pdf = pd.DataFrame(self.native.to_dict(orient="series"))
+        for k, v in new_schema.items():
+            if not v.type.equals(self.schema[k].type):
+                old_type = self.schema[k].type
+                new_type = v.type
+                # int -> str
```
Comment (kvnkho) [inline fugue/dataframe/pandas_dataframe.py:126]: I was wondering why there are only two cases here (similar with Dask DataFrame) with `int to str` and `bool to str`. I assume these are the cases where NULLs can not be casted but other cases work? Thinking specifically about NULL casting as datetime.

**Review by kvnkho — APPROVED** [review]
It looks good. Just wrote some comments but feel free to not address except for the alter_columns docstring. I think this one is not that difficult and it should not be the same as the rename docstring

File: fugue/dataframe/dataframe.py [hunk 2]
```diff
@@ -186,6 +187,17 @@ def rename(self, columns: Dict[str, str]) -> "DataFrame":  # pragma: no cover
         """
         raise NotImplementedError
 
+    @abstractmethod
+    def alter_columns(self, columns: Any) -> "DataFrame":  # pragma: no cover
```
[the remark this answers is not in the document]
Comment (goodwanghan) [reply fugue/dataframe/dataframe.py:191]: Great catch

File: fugue_test/dataframe_suite.py [hunk 3]
```diff
@@ -220,3 +220,114 @@ def test_head(self):
...
+            assert [["a", 1], ["c", None]] == ndf.as_array(type_safe=True)
+            assert ndf.schema == df.schema
+
+            # bool -> str
+            df = self.df([["a", True], ["b", False], ["c", None]], "a:str,b:bool")
+            ndf = df.alter_columns("b:str")
+            actual = ndf.as_array(type_safe=True)
+            expected1 = [["a", "True"], ["b", "False"], ["c", None]]
+            expected2 = [["a", "true"], ["b", "false"], ["c", None]]
+            assert expected1 == actual or expected2 == actual
```
[the remark this answers is not in the document]
Comment (goodwanghan) [reply fugue_test/dataframe_suite.py:251]: spark will convert them to lower case, other frameworks will convert them to upper case

[the remark this answers is not in the document]
Comment (goodwanghan) [reply fugue_test/dataframe_suite.py:251]: added additional comment

File: fugue_test/dataframe_suite.py [hunk 4]
```diff
@@ -220,3 +220,114 @@ def test_head(self):
...
+                [["1", "2020-01-01"], ["2", "2020-01-02 01:02:03"], ["3", None]],
+                "a:str,b:str",
+            )
+            ndf = df.alter_columns("b:datetime,a:int")
+            assert [
+                [1, datetime(2020, 1, 1)],
+                [2, datetime(2020, 1, 2, 1, 2, 3)],
+                [3, None],
+            ] == ndf.as_array(type_safe=True)
+            assert ndf.schema == "a:int,b:datetime"
```
[the remark this answers is not in the document]
Comment (goodwanghan) [reply fugue_test/dataframe_suite.py:333]: we have exception test in `test_get_altered_schema`

[the remark this answers is not in the document]
Comment (goodwanghan) [reply fugue_test/dataframe_suite.py:333]: but what you suggested is very necessary, adding

File: fugue/dataframe/pandas_dataframe.py (hunk 1 above)
Comment (goodwanghan) [reply fugue/dataframe/pandas_dataframe.py:126]: there are also datetime and double, NAT and NAN should be casted to NULLS, and these are all in the unit test suite, so i didn't add them here meaning pandas and dask can handle them or in the new DataFrame init, the enforce type can handle them.

Anyway, the tests are their to ensure they all have expected behaviors

File: fugue_test/dataframe_suite.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (goodwanghan) [reply fugue_test/dataframe_suite.py:335]: It turns out Spark will silently cast invalid types to null ...
But i have added this failed case for all other dataframes

**Pull request merged**